[stdlib] Set, Dictionary: Mix the hash table size into the hash seed

lorentey opened the pull request:
This is done by xoring the bucket count of the hash table to the hash seed. This only flips a single bit in the seed, but avalanche effects make that more than enough to break the correlation between element orderings in tables of different sizes.

This eliminates quadratic performance issues when we copy elements between hash tables of different sizes.

Fixes https://bugs.swift.org/browse/SR-3268

lorentey committed:
[stdlib] Set, Dictionary: Enable per-instance hash seeding

This is done by xoring the base address of the hash table storage to the hash seed.

In deterministic mode, we perturb the seed by the bucket count instead, so that we ordering within hashed collections remains repeatable (as long as the capacity doesn't change).

Fixes https://bugs.swift.org/browse/SR-3268

lorentey commented:
I think this will need adjustments in non-verbatim bridging at minimum. Let's see if tests uncover anything else.

@swift-ci please test

lorentey commented:
@swift-ci please smoke benchmark

natecook1000 commented:
Doesn't using the storage address break copy-on-write?

```swift
let a = [1: "one", 2: "two"]
var b = a
b[2] = "deux"        
```
The last line creates a new block of storage with a different base address, but just copies over the contents. Without rehashing and re-ordering everything, a different perturbation will mean misses.

lorentey commented:
@natecook1000 Oh yeah, oops! COW copies must inherit the original's seed -- not even indices would work otherwise. Luckily, restricting the change to per-capacity seeding would work just fine, because capacity reallocations are allowed to break indices. 🎉

We can reenable per-"instance" seeding later, but we'll need to add the seed as a storage class ivar, and make sure to reuse the same seed when we're making a COW copy.

lorentey committed:
[stdlib] Set, Dictionary: Always use capacity-based seeding

Per-instance seeding would break COW, unless we stored the seed within the storage class.

lorentey commented:
@swift-ci please test

lorentey commented:
@swift-ci please smoke benchmark

lorentey commented:
@swift-ci please test

lorentey commented:
@swift-ci please smoke benchmark

lorentey commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -3989,10 +3989,18 @@ extension _Native${Self}Buffer
+    //
+    // FIXME: Use true per-instance seeding. This requires adding the local seed
+    // into the storage class, because COW copies need to use the same seed as
+    // their original instance.
```
Wow, I can't speaks English good 😳 (I'll fix it when the tests are done.)

lorentey commented:
This seems to be a performance win, and not just for the known quadratic cases (DictionaryFilter and -Copy). ✨

I did not expect to see those two regressions. They're not too bad, it's just weird that they exist.

lorentey committed:
Clarify rationale for using a per-capacity hash seed

lorentey committed:
[test] Test that hashed collections use local hash seeds

natecook1000 commented on validation-test/stdlib/Dictionary.swift:
```diff
@@ -4574,6 +4574,37 @@ DictionaryTestSuite.test("removeAt") {
+  }
+  // If this test fails, there is a problem with local hash seeding.
+  expectFalse(small.map{$0.key}.elementsEqual(slice.map{$0.key}))
+}
```
This is a super cool test 👏👏

lorentey replied:
Collision chains at the 1% mark will still make this sometimes pass without correct seeding. ;(

Luckily, I think we can fix this by simply ignoring the first "few" copied elements.

natecook1000 replied:
I think we could also reduce the number of chains by allocating more for the big dictionary, too—if you start with `[Int: Int](minimumCapacity: count * 3)` or something there should be enough space that the collision chains are rare/short.

lorentey committed:
[test] Fix ~10% false pass rate without local hash seeds

lorentey commented:
@swift-ci please test

lorentey commented:
@swift-ci please benchmark

lorentey commented:
@swift-ci please test linux platform

lorentey commented:
@swift-ci benchmark

lorentey commented:
@swift-ci please test

lorentey commented:
@please benchmark

lorentey commented:
@please benchmark

lorentey commented:
@swift-ci please benchmark

lorentey commented:
@swift-ci please benchmark

lorentey commented:
@swift-ci test

lorentey commented:
@swift-ci test

lorentey commented:
@swift-ci benchmark

lorentey commented:
@swift-ci benchmark

lorentey commented:
Let's try again with https://github.com/apple/swift/pull/15300

@swift-ci benchmark

lorentey commented:
Fifty-third time is the charm! 🍀

@swift-ci benchmark

lorentey committed:
[stdlib] Set/Dictionary: Cache the hash seed in the storage instance

This is a significant performance win for small tables.

lorentey commented:
@swift-ci test

lorentey commented:
@swift-ci test

lorentey commented:
@swift-ci benchmark

lorentey commented:
@swift-ci benchmark

lorentey commented:
@swift-ci benchmark

lorentey commented:
@swift-ci benchmark

lorentey commented:
@swift-ci benchmark

lorentey commented:
Wow, these benchmark results are noisy. `DictionaryFilter` and `DictionaryCopy` may be the only meaningful results here.

* `Sim2DArray`/`Hanoi`/`CharIteration*`/`CharIndexing*`/`RangeIterationSigned` don't use dictionaries
* `WordSplitASCII`/`WordSplitUTF16` is known to be flaky at the moment (#15337)
* `Histogram` seems to be sensitive to the random hash seed (I'll submit a PR to disable it while benchmarking for now; a full fix needs an update to `Dictionary`'s storage representation.)

LGTM! 🎉
